### ISSUE TITLE
Shairport-Sync using Airfoil -- delay between AirPlay-Speakers and version 3.1.7 [resolved]

### ISSUE THREAD

**Issue by LuckyLukeCH** [mikebrady/shairport-sync#639]
Hello Mike,

First of all, thank you very much for your huge work for this very nice project!
After finally figuring out the equalizer-option, I'm installing it in different rooms in our house.

No I set up a new shairport-sync player version 3.1.7 and interesting got a negative (earlier) delay to the older version 2.8.4 resp. the official AirPlay-Speakers.
What I found in the forum, is that there was a change in iTunes from 2 to 2.25s. It could be around this difference.
The delay only appears when playing through AirFoil from Mac; directly from iTunes all speakers are in sync!
Can you give me a hint what to adjust or test to remove this delay?

Thank you very much and regards
Lukas Gaam

**Comment by mikebrady** [comment]
Thanks for the post, Lukas. 

I'd like just to clarify your scenario please. Is the following description correct?

With Shairport Sync 3.1.7:
1. From a Mac using iTunes, the speakers on the Mac and all Shairport-Sync-powered speakers are in sync.
2. From the same Mac using AirFoil, the speakers on the Mac and the Shairport-Sync-powered speakers are not in sync.

Please feels free to correct it it it's wrong. Also, could you tell me what version of AirFoil you are using please? And are the Shairport-Sync-powered speakers ahead or behind the Mac's speakers?

**Comment by LuckyLukeCH** [comment]
Thank you for your fast response, Mike!

1. Yes, with iTunes all AirPlay-Speakers and Shairport-Sync speakers 2.8.4 and 3.1.7 are IN sync.
2. Yes, with AirFoil the AirPlay-Speakers and Shairport-Sync speakers version 3.1.7 are NOT sync.
Only the Shairport-Sync speakers version 2.8.4 are in sync with the AirPlay-Speakers

- AirFoil is version 5.6.4
- The Shairport-Sync speakers version 3.1.7 are ahead the others, could be around 250ms...

Thank you and regards
Lukas Gaam

**Comment by mikebrady** [comment]
Thanks. Let me look into it...

**Comment by mikebrady** [comment]
Well, I can reproduce the problem, so let me check the thing out.

**Comment by mikebrady** [comment]
Okay, I've posted a new development version with a fix to the problem, I hope. I'd be grateful if you would try it out. Also, I changed the title of this post a little bit -- I hope you don't mind.

**Comment by LuckyLukeCH** [comment]
Hello Mike,

Thank you very much for your fast effort! I hope to find the time in the next days to test it and will post the result.

The title is perfect now - don't worry ;-)

Regards
Lukas

**Comment by LuckyLukeCH** [comment]
Hello Mike,

I was not successful yet; the delay seems still the same. 
Could you please confirm the compilation process:

`cd /root && git clone https://github.com/mikebrady/shairport-sync.git
cd /root/shairport-sync && git checkout -b development
autoreconf -i -f /root/shairport-sync
./configure --sysconfdir=/etc --with-alsa --with-avahi --with-ssl=openssl --with-metadata  --with-systemd
make && make install
`

Thank you and regards
Lukas

**Comment by mikebrady** [comment]
Hi Lukas. Remember to restart Shairport Sync afterwards, with `$ sudo systemctl restart shairport-sync`. Otherwise, the old version of Shairport Sync will continue to be used.

Meanwhile, I will try your exact command sequence...

**Comment by mikebrady** [comment]
There are mistakes in your command sequence. Please wait...

**Comment by mikebrady** [comment]
Okay, there were some mistakes in your command sequence.

Beforehand, though (and this is not the cause of your problem) as a matter of good practice and security, you should not work like this when logged in as the `root` superuser. Instead you should do as much as you can as a regular unprivileged user (who has `sudo` privileges), and only go into a `root` mode when necessary, using `sudo`.

The following is a transcript of a normal user "mike" (with `sudo` privileges) starting in the home directory. You'll see the use of `sudo` at the very end to install and restart the program. You may be asked to enter your password.

Here is the full command (starting at `cd `) but with the command line prompt:
```
mike@ubuntu:~/$ cd ~ && git clone https://github.com/mikebrady/shairport-sync.git && cd shairport-sync && git checkout development && autoreconf -if && ./configure --sysconfdir=/etc --with-alsa --with-avahi --with-ssl=openssl --with-metadata --with-systemd && make && sudo make install && sudo systemctl restart shairport-sync
```
Here is a log of the activity:

```
Cloning into 'shairport-sync'...
remote: Counting objects: 8296, done.
remote: Compressing objects: 100% (116/116), done.
remote: Total 8296 (delta 103), reused 91 (delta 52), pack-reused 8128
Receiving objects: 100% (8296/8296), 3.96 MiB | 4.27 MiB/s, done.
Resolving deltas: 100% (5414/5414), done.
Branch development set up to track remote branch development from origin.
Switched to a new branch 'development'
configure.ac:20: installing './compile'
configure.ac:6: installing './install-sh'
configure.ac:6: installing './missing'
Makefile.am: installing './INSTALL'
Makefile.am: installing './depcomp'
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking dependency style of g++... gcc3
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for clock_gettime in -lrt... yes
>>Including libdaemon
checking for DAEMON... yes
checking for pthread_create in -lpthread... yes
checking for exp in -lm... yes
>>Including libpopt
checking for POPT... yes
checking for LIBCONFIG... yes
checking encryption libraries chosen... checking for SSL... yes
>>Including metadata support
>>Including Avahi mDNS support
checking for avahi_client_new in -lavahi-client... yes
checking for avahi_strerror in -lavahi-common... yes
>>Including an ALSA back end
checking for ALSA... yes
checking for xmltoman... xmltoman
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking getopt_long.h usability... no
checking getopt_long.h presence... no
checking for getopt_long.h... no
checking arpa/inet.h usability... yes
checking arpa/inet.h presence... yes
checking for arpa/inet.h... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking limits.h usability... yes
checking limits.h presence... yes
checking for limits.h... yes
checking mach/mach.h usability... no
checking mach/mach.h presence... no
checking for mach/mach.h... no
checking for memory.h... (cached) yes
checking netdb.h usability... yes
checking netdb.h presence... yes
checking for netdb.h... yes
checking netinet/in.h usability... yes
checking netinet/in.h presence... yes
checking for netinet/in.h... yes
checking for stdint.h... (cached) yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking sys/ioctl.h usability... yes
checking sys/ioctl.h presence... yes
checking for sys/ioctl.h... yes
checking sys/socket.h usability... yes
checking sys/socket.h presence... yes
checking for sys/socket.h... yes
checking sys/time.h usability... yes
checking sys/time.h presence... yes
checking for sys/time.h... yes
checking syslog.h usability... yes
checking syslog.h presence... yes
checking for syslog.h... yes
checking for unistd.h... (cached) yes
checking for inline... inline
checking for int16_t... yes
checking for int32_t... yes
checking for int64_t... yes
checking for pid_t... yes
checking for size_t... yes
checking for ssize_t... yes
checking for uint16_t... yes
checking for uint32_t... yes
checking for uint64_t... yes
checking for uint8_t... yes
checking for working alloca.h... yes
checking for alloca... yes
checking for error_at_line... yes
checking vfork.h usability... no
checking vfork.h presence... no
checking for vfork.h... no
checking for fork... yes
checking for vfork... yes
checking for working fork... yes
checking for working vfork... (cached) yes
checking for atexit... yes
checking for clock_gettime... yes
checking for gethostname... yes
checking for inet_ntoa... yes
checking for memchr... yes
checking for memmove... yes
checking for memset... yes
checking for mkfifo... yes
checking for pow... yes
checking for select... yes
checking for socket... yes
checking for stpcpy... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strdup... yes
checking for strerror... yes
checking for strstr... yes
checking for strtol... yes
checking for strtoul... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating man/Makefile
config.status: creating scripts/shairport-sync.service
config.status: creating scripts/shairport-sync
config.status: creating config.h
config.status: executing depfiles commands
make  all-recursive
make[1]: Entering directory '/home/mike/shairport-sync'
Making all in man
make[2]: Entering directory '/home/mike/shairport-sync/man'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/home/mike/shairport-sync/man'
make[2]: Entering directory '/home/mike/shairport-sync'
depbase=`echo shairport.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT shairport.o -MD -MP -MF $depbase.Tpo -c -o shairport.o shairport.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo rtsp.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT rtsp.o -MD -MP -MF $depbase.Tpo -c -o rtsp.o rtsp.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo mdns.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT mdns.o -MD -MP -MF $depbase.Tpo -c -o mdns.o mdns.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo mdns_external.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT mdns_external.o -MD -MP -MF $depbase.Tpo -c -o mdns_external.o mdns_external.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo common.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT common.o -MD -MP -MF $depbase.Tpo -c -o common.o common.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo rtp.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT rtp.o -MD -MP -MF $depbase.Tpo -c -o rtp.o rtp.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo player.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT player.o -MD -MP -MF $depbase.Tpo -c -o player.o player.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo alac.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT alac.o -MD -MP -MF $depbase.Tpo -c -o alac.o alac.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo audio.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT audio.o -MD -MP -MF $depbase.Tpo -c -o audio.o audio.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo loudness.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT loudness.o -MD -MP -MF $depbase.Tpo -c -o loudness.o loudness.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo mdns_avahi.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT mdns_avahi.o -MD -MP -MF $depbase.Tpo -c -o mdns_avahi.o mdns_avahi.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo audio_alsa.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -DHAVE_CONFIG_H -I.  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -Wno-multichar -DSYSCONFDIR=\"/etc\"  -g -O2 -MT audio_alsa.o -MD -MP -MF $depbase.Tpo -c -o audio_alsa.o audio_alsa.c &&\
mv -f $depbase.Tpo $depbase.Po
g++  -g -O2   -o shairport-sync shairport.o rtsp.o mdns.o mdns_external.o common.o rtp.o player.o alac.o audio.o loudness.o  mdns_avahi.o  audio_alsa.o               -lasound -lavahi-common -lavahi-client -lssl -lcrypto -lconfig -lpopt -lm -lpthread -ldaemon -lrt 
make[2]: Leaving directory '/home/mike/shairport-sync'
make[1]: Leaving directory '/home/mike/shairport-sync'
make  install-recursive
make[1]: Entering directory '/home/mike/shairport-sync'
Making install in man
make[2]: Entering directory '/home/mike/shairport-sync/man'
make[3]: Entering directory '/home/mike/shairport-sync/man'
make[3]: Nothing to be done for 'install-exec-am'.
 /bin/mkdir -p '/usr/local/share/man/man7'
 /usr/bin/install -c -m 644 shairport-sync.7 '/usr/local/share/man/man7'
make[3]: Leaving directory '/home/mike/shairport-sync/man'
make[2]: Leaving directory '/home/mike/shairport-sync/man'
make[2]: Entering directory '/home/mike/shairport-sync'
make[3]: Entering directory '/home/mike/shairport-sync'
 /bin/mkdir -p '/usr/local/bin'
  /usr/bin/install -c shairport-sync '/usr/local/bin'
make  install-exec-hook
make[4]: Entering directory '/home/mike/shairport-sync'
[ -e /etc ] || mkdir /etc
cp scripts/shairport-sync.conf /etc/shairport-sync.conf.sample
[ -f /etc/shairport-sync.conf ] || cp scripts/shairport-sync.conf /etc/shairport-sync.conf
getent group shairport-sync &>/dev/null || groupadd -r shairport-sync >/dev/null
getent passwd shairport-sync &> /dev/null || useradd -r -M -g shairport-sync -s /usr/bin/nologin -G audio shairport-sync >/dev/null
[ -e /lib/systemd/system ] || mkdir -p /lib/systemd/system
[ -f /lib/systemd/system/shairport-sync.service ] || cp scripts/shairport-sync.service /lib/systemd/system
make[4]: Leaving directory '/home/mike/shairport-sync'
make[3]: Nothing to be done for 'install-data-am'.
make[3]: Leaving directory '/home/mike/shairport-sync'
make[2]: Leaving directory '/home/mike/shairport-sync'
make[1]: Leaving directory '/home/mike/shairport-sync'
mike@ubuntu:~$
```
Finally, to ensure the correct version has been installed, I checked its version as follows:
```
shairport-sync -V
3.2d22-OpenSSL-Avahi-ALSA-metadata-sysconfdir:/etc
mike@ubuntu:~/$

**Comment by mikebrady** [comment]
Hi there. Did you have any joy with this?

**Comment by kariagekun** [comment]
I can confirm this, I was having similar issue with version 3.1.7. 

After downloading the development version 3.2d26-OpenSSL-Avahi-ALSA-metadata-sysconfdir:/etc, using the command above, it seems to fix the delay problem with Airfoil Version 5.6.5.

Thanks!

**Comment by mikebrady** [comment]
Thanks for this contribution. Closing the issue. Please open a new one if necessary.
